lib/app/modules/us_core_r4/clinicalnotes_sequence.rb

```diff
@@ -0,0 +1,211 @@
+# frozen_string_literal: true
+
+module Inferno
+  module Sequence
+    class UsCoreR4ClinicalNotesSequence < SequenceBase
+      group 'US Core R4 Profile Conformance'
+
+      title 'Clinical Notes Guideline Tests'
+
+      description 'Verify that DocumentReference and DiagnosticReport resources on the FHIR server follow the US Core R4 Clinical Notes Guideline'
+
+      test_id_prefix 'ClinicalNotes'
+
+      requires :token, :patient_id
+      conformance_supports :DocumentReference, :DiagnosticReport
+
+      details %(
+
+        The #{title} Sequence tests DiagnosticReport and DocumentReference resources associated with the provided patient.  The resources
+        returned will be checked for consistency against the [US Core Clinical Notes Guidance](https://build.fhir.org/ig/HL7/US-Core-R4/clinical-notes-guidance.html)
+
+      )
+
+      @clinicalnotes_found = false
+
+      def test_clinical_notes_document_reference(type_code)
+        skip 'No Clinical Notes appear to be available for this patient. Please use patients with more information.' unless @clinicalnotes_found
+
+        assert @actual_type_codes.include?(type_code), "Clinical Notes shall have at least one DocumentReference with type #{type_code}"
+      end
+
+      def test_clinical_notes_diagnostic_report(category_code)
+        skip 'No Clinical Notes appear to be available for this patient. Please use patients with more information.' unless @clinicalnotes_found
+
+        search_params = { 'patient': @instance.patient_id, 'category': category_code }
+
+        reply = get_resource_by_params(versioned_resource_class('DiagnosticReport'), search_params)
+        assert_response_ok(reply)
+        assert_bundle_response(reply)
+
+        resource_count = reply&.resource&.entry&.length || 0
+        @resources_found = true if resource_count.positive?
+
+        assert @resources_found, "Clinical Notes shall have at least one DiagnosticReport with category #{category_code}"
+
+        diagnostic_reports = reply&.resource&.entry&.map { |entry| entry&.resource }
+
+        diagnostic_reports&.each do |diarpt|
+          diarpt&.presentedForm&.select { |attachment| attachment&.url&.present? }&.each do |attachment|
+            assert @attachment_urls.key?(attachment.url), "Attachment #{attachment.url} referenced in DiagnosticReport/#{diarpt.id} but not in any DocumentReference"
```

Comment:
I see two issues with `@attachment_urls`:
- the name `flag` doesn't say what is being flagged
- it's called `@attachment_urls` but it's really just the attachment urls from docrefs

I'd address both of these by replacing `@attachment_urls` with two instance variables, `@docref_attachments` and `@report_attachments`. These variables could be hashes where the keys are attachment urls, and the values are the resource id `@docref_attachments[attachment.url] = docref.id`. Then this assertion could be removed from this test, and test 10 could do this for both sets of urls. Subtracting arrays gives their difference, so you can easily get a list of missing attachments.
```
irb(main):006:0> @docref_urls = ['url1', 'url2', 'url3', 'url4']
=> ["url1", "url2", "url3", "url4"]
irb(main):007:0> @report_urls = ['url3', 'url1']
=> ["url3", "url1"]
irb(main):008:0> @docref_urls - @report_urls
=> ["url2", "url4"]
```

Reply:
Add Class ClinicalNoteAttachment to save resource_type and attachment hash.
Add function assert_attachment_matched to compare two attachment hashes and build error message